Add emacs config file and references to README.md (#46)
kth-step/HolBA: CONTRIBUTING.md

# Coding style

* like HOL source code
  - Spaces, no tabs
  - No unicode
  - `snake_case` (e.g. `bir_number_of_mem_splits_def`)

# Branch policy

## `master` branch

`master` is the branch where every feature is available, but not necessarily finalized:

>  - Can cheat, but has to be avoided (allowed only in a small number of modules, which are explicitly marked unstable, no-cheat CI test mentioned below would need a corresponding exception list)

Reviewer: I didn't find anything about **no-cheat** CI tests specifically. Did you forget it?

Author: no, but it wasn't as wordy as you expected it possibly :)

 - Code should not be be commented out
 - Our Travis CI must pass
   - **Holmake must work** (i.e. must correctly compile)
   - All tests must succeed
 - Bug-fixing commits are ok
 - At least 1 accepting review is needed in order to merge into `master`

Notice:
 - **No development happens on the `master` branch**, but rather on separate feature branches
 - **In order to prevent mayhem**, define good interfaces for your code (so that development won't break existing code)

Follow these instructions whenever you merge to master:

>   - `grep` for "cheat" (TODO: add this as a CI test)

Reviewer: Prefer a new issue instead of this TODO that we will forget.

Author: done, and kept as TODOs. pushing now

  - Check that the `README` is up to date (especially tool status)
  - Find a reviewer for your Pull Request

## tags

tags are like `master` and on top of this:
 - Should have as many **completed features** as possible
 - No cheats in the directory `src`

## Feature branches

Every "somewhat" working tool should be available in the `master` branch, but new
features or any development must go on new branches prefixed with `dev_`, fixes with `fix_`.

Guidelines:
 - Branch names must be short and explicit (prefer explicit over short)
 - Every feature branch should involve small developments
 - Rebase feature branches on `master` **often**, by using `git rebase` or `git merge`

>  - **Merge feature branches to `master` often**: work on small features

Reviewer: So, should we merge or rebase often?

Author: I wonder how these two similar bullet points got there. It was just moved from the README.md file originally.

## Coding style

* HOL source code
  - Spaces, no tabs
  - No unicode
  - `snake_case` (e.g. `bir_number_of_mem_splits_def`)

## Branch policy

### tags

tags should have as many **completed features** as possible:

>  - no cheat

Reviewer: **See next comment about cheat in `master` before.**

Is that possible? If you tag on master, where development happens, then maybe there is some `cheat`?

Author: let's start with the good intentions and then we can gradually go back in case it is hard to keep them up. in case it is hard to follow up on this, we could tag instable modules or directories, where tests are not required to succeed (for example)

Reviewer: :+1: That's what I had in mind: tag and mention tools status.

Author: let's build the support for this later and just mention it here for now. so we settle for a tagging of unstable directories which enables usage of cheat and failing tests in there

Author: and we build it as soon as we need it

 - must correctly compile

>  - self tests must succeed

Reviewer: ```suggestion
 - all tests must succeed, which is checked by Travis CI
```

 - code should be tested

Follow these instructions whenever you merge to master:

>   - `grep` for "cheat"

Reviewer: Is that still true, now that master is our dev branch, and that we should merge often?

Author: let's fix a test that checks exactly this, and i would say yes

Reviewer: How do you want to check that? Is there a wait to forbid cheating with Holmake?

Author: no, but it could grep cheat 😄

Reviewer: Yeah but that doesn't work, just try it on your local repository. However, you can grep the build logs, Holmake writes "CHEATED" when that's the case.

Author: then this means grepping sml files only

  - check that the `README` is up to date (especially tool status)
  - find a reviewer for your Pull Request

> ### `master` branch

Reviewer: This title should be before `Follow these instructions whenever you merge to master:`

Author: yes

`master` is the branch where every feature is available, but not necessarily finalized:

>   - Can cheat, but has to be avoided

Reviewer: Contradicts what is said above.

Author: that is true. i think we wrote this here together and didn't fix the other one. and this is also the reason why you asked for the other one now. until now we didn't need cheat in there, so let's see how long we can keep it like this.

  - Code should not be be commented out

>   - **Holmake must work**, which is checked by our Travis CI

Reviewer: Should we add that all tests must pass? That eventually shouldn't be the case, but we cannot really define failing tests for the time being. An alternative would be to always manually check the failing tests in Travis.

Author: good point, let's just keep it in mind for now and see how it goes

>   - bug-fixes are ok

Reviewer: ```suggestion
  - bug-fixing commits are ok
```
Is that what you mean?

>   - 1 review is needed in order to merge into `master`

Reviewer: ```suggestion
  - at least 1 review is needed in order to merge into `master`
```

> However, **no development happens on this branch**, but rather on separate

Reviewer: ```suggestion
However, **no development happens on the `master` branch**, but rather on separate
```
